### PR TITLE
Better type checks for NonNull and List

### PR DESCRIPTION
**Pull request by sungam3r**

### CODE REVIEW

File: src/GraphQL/Types/Composite/ListGraphType.cs [hunk 1]
```diff
@@ -27,23 +16,48 @@ public ListGraphType(IGraphType type)
...
+        /// Initializes a new instance for the specified inner graph type.
+        /// </summary>
+        public ListGraphType()
+            : base(null)
+        {
+            _type = typeof(T);
+        }
+
+        /// <inheritdoc/>
+        public override Type Type => _type;
```
Comment (Shane32) [inline src/GraphQL/Types/Composite/ListGraphType.cs:61]: `Type` is only used when the schema is built, correct?  There should be no reason to cache the value.
```suggestion
        public override Type Type => typeof(T);
```

File: src/GraphQL/Types/Composite/NonNullGraphType.cs [hunk 2]
```diff
@@ -24,37 +13,62 @@ public class NonNullGraphType : GraphType, IProvideResolvedType
...
+            if (typeof(NonNullGraphType).IsAssignableFrom(typeof(T)))
+            {
+                throw new ArgumentOutOfRangeException("type", "Cannot nest NonNull inside NonNull.");
+            }
+
+            _type = typeof(T);
+        }
+
+        /// <inheritdoc/>
+        public override Type Type => _type;
```
Comment (Shane32) [inline src/GraphQL/Types/Composite/NonNullGraphType.cs:72]:
```suggestion
        public override Type Type => typeof(T);
```
See above.  Also need to remove `_type` if this is changed.

File: src/GraphQL/Types/Composite/ListGraphType.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/Types/Composite/ListGraphType.cs:61]: 🤦

**Comment by Shane32** [comment]
We should make it standard practice to write a test for added checks.  It would only take a few lines a code.  I would not check for the exact error message but simply that `ArgumentOutOfRangeException` occurred.

File: src/GraphQL/Types/Composite/NonNullGraphType.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/Types/Composite/NonNullGraphType.cs]: Also 🤦

**Comment by sungam3r** [comment]
OK. Red Codecov report motivates me to do it now.

**Review by Shane32 — APPROVED** [review]

**Pull request merged**